lib/hipe/icode/hipe_beam_to_icode.erl

```diff
@@ -222,25 +222,32 @@ leafness([I|Is], Leafness) ->
 	  end;
 	T when T =:= 'local' orelse T =:= 'remote' ->
 	  {M,F,A} = hipe_icode:enter_fun(I),
-	  case erlang:is_builtin(M, F, A) of
-	    true -> leafness(Is, Leafness);
+	  case erlang:is_builtin(M, F, A)of
+	    true -> leafness(Is, IsClosure, Leafness);
+	    _ when IsClosure -> leafness(Is, IsClosure, closure);
 	    _ -> false
 	  end
       end;
-    _ -> leafness(Is, Leafness)
+    _ -> leafness(Is, IsClosure, Leafness)
   end.
 
 %% XXX: this old stuff is passed around but essentially unused
 is_leaf_code(Leafness) ->
   case Leafness of
     true -> true;
     selfrec -> true;
+    closure -> false;
     false -> false
   end.
 
 needs_redtest(Leafness) ->
   case Leafness of
     true -> false;
+    %% A "leaf" closure may contain tailcalls to non-closures in addition to
+    %% what othere leaves may contain. Omitting the redtest is safe since they
```

Comment:
othere -> other (in this line)
"be" missing in the next one (or something else is wrong in the phrasing)